server/src/main/java/org/apache/druid/metadata/SqlSegmentsMetadataQuery.java

```diff
@@ -312,53 +327,91 @@ public int markSegments(final Collection<SegmentId> segmentIds, final boolean us
   }
 
   /**
-   * Marks all used segments that are *fully contained by* a particular interval as unused.
+   * Marks all used segments that are <b>fully contained by</b> a particular interval as unused.
    *
-   * @return the number of segments actually modified.
+   * @return Number of segments updated.
    */
   public int markSegmentsUnused(final String dataSource, final Interval interval)
+  {
+    return markSegmentsUnused(dataSource, interval, null);
+  }
+
+  /**
+   * Marks all used segments that are <b>fully contained by</b> a particular interval filtered by an optional list of versions
+   * as unused.
+   *
+   * @return Number of segments updated.
+   */
+  public int markSegmentsUnused(final String dataSource, final Interval interval, @Nullable final List<String> versions)
   {
     if (Intervals.isEternity(interval)) {
-      return handle
-          .createStatement(
-              StringUtils.format(
-                  "UPDATE %s SET used=:used, used_status_last_updated = :used_status_last_updated "
-                  + "WHERE dataSource = :dataSource AND used = true",
-                  dbTables.getSegmentsTable()
-              )
+      final StringBuilder sb = new StringBuilder();
+      sb.append(
+          StringUtils.format(
+              "UPDATE %s SET used=:used, used_status_last_updated = :used_status_last_updated "
+              + "WHERE dataSource = :dataSource AND used = true",
+              dbTables.getSegmentsTable()
           )
+      );
+
+      final boolean hasVersions = !CollectionUtils.isNullOrEmpty(versions);
```

Comment:
This doesn't seem right. The behaviour for null should not be the same as empty.
The calling code might be filtering out the given list of versions based on some condition, resulting in an empty list of versions being passed down to this method. In such a case, this method should ideally return no results instead of all versions.

If this was already the case in the existing code, we should consider fixing it or atleast clearly calling this out in the javadocs of the whole chain of methods that lead from the rest API endpoint down to this method.

PS: Treating null and empty as the same probably makes sense at the API level, since the API would be clearly documented and would give out clear validation messages such as `Versions must be specified` or `Versions not required when specifying segment IDs`.

Reply:
Yeah, good point. It makes sense to treat them differently at the DB level since it's also used by other internal calls, not just by the user-facing APIs. I quickly checked the usages, and it's either null or a non-empty set, so we should be good as-is. I will address your comment in a follow-up change and call out any difference in the javadocs as needed.